### ISSUE TITLE
GCI 2014 asking for a favor

### ISSUE THREAD

**Issue by amroto** [amber-smalltalk/amber#1123]
Now I am working on a task in google code in to work on the tutorial of amber but it seems to be out of date so can somebody guide me how to make the counter and add it to my HTML and run it and also everytime I add a class and refresh its gone and almost all the links are out of date, so as a new user to amber can somebody guide me with creating the counter and running it on 0.13 amber.

**Comment by amroto** [comment]
https://github.com/amroto/Amber-Tutorial

**Comment by amroto** [comment]
please give me feedback

**Comment by herby** [comment]
I think I have already written that changing require load list in index.html is wrong, so please fix it it so you do not touch index,html at all and use deploy.js instead, as it was designed to.

**Comment by amroto** [comment]
I didnt understand how to do that ! 
As you see I am completely new to programming so this is a bit tough to me

**Comment by herby** [comment]
Adding a line into deploy.js (into a place which is even marked with 'put package to be loaded here' comments) is IMO less tough than updating array in the require call inside index.html. Not to mention this is how it must be done this way for all the other 'batteries included' (deployment of an app) to work. If you do a specific project which lies outside the boundaries of the standard approach, you can of course load amber in other ways. But for 95% of the SPA projects, you should just go with the standard and use deploy.js to list app packages. And not touching the `require(['app']` part in index.html.

amroto wrote:

**Comment by hhzl** [comment]
@amroto I have opened an issue to do this [here](https://github.com/amber-smalltalk/amber-examples/issues/72). Let's continue the discussion there.

**Comment by amroto** [comment]
I was able to load the package without the need to edit the HTML file but making the counter itself load without running doit from helios was a failure,so that I can update the tutorial, any advice?
and also I want to create a web page with a list (any kind as colors , names ..etc) but I wanted to be edittable and being able to add and delete data from the table/list as I want any help on that as well ?

**Comment by amroto** [comment]
@herby

**Comment by herby** [comment]
I am pretty confused from "but making the counter itself load without running doit from helios was a failure,so that I can update the tutorial". I just don't understand :-\ Can you rephrase somehow?

**Comment by amroto** [comment]
https://github.com/amroto/Amber-Tutorial/blob/master/README.md

**Comment by amroto** [comment]
look at the tutorial
I removed the part where I edit require(['app''], function (amber) {
amber.initialize({
'transport.defaultAmdNamespace': "amber-myapp"
});

**Comment by amroto** [comment]
but do I still have to add:
amber.globals.TCounter._run();
so that the counter loads automatically or is there a new/other way ?

Plus can you help me create a webpage which loads a list , where this list is completely editable (like I can add/remove info )

**Comment by amroto** [comment]
@herby

**Comment by amroto** [comment]
and how to update to amber 0.14

**Comment by amroto** [comment]
the list I was talkin about is something similar to this : 
http://ngtg.techgrind.asia:7999/user/#/user

**Comment by herby** [comment]
Running Counter on startup: why not to use MyApp which is started anyway and just add it to start method? IIRC it calls `self augmentPage` but it can do lots of other things you want to do on app startup. it's your app, after all.

**Comment by herby** [comment]
As for updating amber to 0.14.0:

```
bower uninstall amber --save
bower install amber --save
npm uninstall amber-dev --save
npm install amber-dev --save
```

in the root directory of the project.

**Comment by amroto** [comment]
1- how to make it call a self argument page?
2- And can you guide me on how to create a page smimilar to this http://ngtg.techgrind.asia:7999/user/#/user?

@herby

**Comment by amroto** [comment]
@herby

**Comment by herby** [comment]
amroto wrote:
> 1- how to make it call a self argument page?

???

> 2- And can you guide me on how to create a page smimilar to this
> http://ngtg.techgrind.asia:7999/user/#/user?

I don't know. Copy it and reuse it? Amber is good with reusing JS 
machinery of many kinds.

**Comment by sebastianconcept** [comment]
@amroto you could also use [flow](https://github.com/flow-stack/flow) to do something like that

**Comment by amroto** [comment]
1- how to make it call a self argument page?
2- And can you guide me on how to create a page smimilar to this http://ngtg.techgrind.asia:7999/user/#/user?

@philippeback

**Comment by hhzl** [comment]
@amroto  Did you manage to get the run Counter code?

https://github.com/amber-smalltalk/amber/wiki/The-counter-example

Or is this still an open issue?

**Comment by hhzl** [comment]
Amber has a class ``Counter`` in the ``Examples`` package.  So as a start you may just call that class instead of creating your own counter.

``amber init`` on the command line creates a project template. This includes a file ``index.html`` preconfigured and fully functional.

So the only thing you have to do is to replace the call to the class stub generated by ``amber init``  with a call to the existing counter class. (one line)

Below is a copy of the file ``index.html`` with the fix included.


    <!DOCTYPE html>
    <html>
    
      <head>
        <title>Amber Counter</title>
        <meta http-equiv="content-type" content="text/html; charset=utf-8" />
        <meta name="author" content="hhzl" />
        <script type='text/javascript' src='the.js'></script>
      </head>
    
      <body>
      <p>Hi, hhzl! Welcome to Amber project: "Amber Counter".</p>
      <p>This is the place for your application's HTML. After getting familiar with Amber,
          just remove this welcome contents from index.html and replace it with your own.</p>
      <button id="amber-with">Hello from TagBrush >> with:</button>
      <button id="jquery-append">Hello from jQuery append</button>
      <ol id="output-list"></ol>
      <script type='text/javascript'>
          require(['app'], function (amber) {
              amber.initialize({
                //used for all new packages in IDE
                'transport.defaultAmdNamespace': "amber-ambercounter"
              });
              require(["amber-ide-starter-dialog"], function (dlg) { dlg.start(); });
              // amber.globals.AmberCounter._start();  // generated by amber init
              amber.globals.Counter._tryExample(); // added by Hannes H. to call the
                                                   // existing Counter class method
          });
      </script>
      </body>
    
    </html>

So the counter works out of the box. 

![counter_class_in_amber_smalltalk_2015-01-06 19 42 11](https://cloud.githubusercontent.com/assets/1112023/5634935/37977a40-95dc-11e4-97a9-96031ae6f051.png)


As a second step you may create your own counter class in the package generated by ``amber init``.

**Comment by herby** [comment]
@hhzl What I was telling is that you could just let `amber.globals.AmberCounter._start();` as is in `index.html

**Comment by hhzl** [comment]
So the better solution is not to change the generated ``index.html``.

    <!DOCTYPE html>
    <html>
    
      <head>
        <title>Amber Counter</title>
        <meta http-equiv="content-type" content="text/html; charset=utf-8" />
        <meta name="author" content="hhzl" />
        <script type='text/javascript' src='the.js'></script>
      </head>
    
      <body>
      <p>Hi, hhzl! Welcome to Amber project: "Amber Counter".</p>
      <p>This is the place for your application's HTML. After getting familiar with Amber,
          just remove this welcome contents from index.html and replace it with your own.</p>
      <button id="amber-with">Hello from TagBrush >> with:</button>
      <button id="jquery-append">Hello from jQuery append</button>
      <ol id="output-list"></ol>
      <script type='text/javascript'>
          require(['app'], function (amber) {
              amber.initialize({
                //used for all new packages in IDE
                'transport.defaultAmdNamespace': "amber-ambercounter"
              });
              require(["amber-ide-starter-dialog"], function (dlg) { dlg.start(); });
              amber.globals.AmberCounter._start();  
          });
      </script>
      </body>
    
    </html>

But change the method ``augmentPage`` (instance method side of the ``amber init`` generated class ``AmberCounter``)

from 

    augmentPage
    	'#amber-with' asJQuery click: [ self doAmberWith ].
    	'#jquery-append' asJQuery click: [ self doJQueryAppend ].

to 
    
    augmentPage
        	'#amber-with' asJQuery click: [ self doAmberWith ].
    	        '#jquery-append' asJQuery click: [ self doJQueryAppend ].
	
    	        Counter new appendToJQuery: 'body' asJQuery

**Comment by hhzl** [comment]
@amroto 

There is a ToDo list application for Amber version 0.12 which is well worked out.

The app has three classes
I suggest that you have a look at the GUI class and try to come up with a variant of that GUI.

I started a wiki page which might lead to a tutorial on this just now

https://github.com/amber-smalltalk/amber/wiki/ToDo-list-application

## ``#renderOn:``

The ``#renderOn:`` method for the GUI there is

From: https://github.com/hhzl/Amber-ToDo-List/edit/master/src/Examples-ToDo.st

    !Todo methodsFor: 'rendering'!

    renderOn: html
       | input delete |
       html li
         id: self id;
              with: [
                   input := html input
                   type: 'checkbox';
                       onClick: [
                       self toggleIsDone.
                          onTodoChanged value: self. 
                          (input asJQuery next) toggleClass: 'done' ].
                          html span with: text .
                          delete := html span
                          with: '  [x]';
                       onClick: [ onDeleteClicked value: self ]].
              isDone ifTrue: [ 
                     input at: 'checked' put: 'checked'.
              (input asJQuery next) addClass: 'done'].
    ! !

Use something like this for your own first attempt at creating the GUI you want.

If you have a codeing question you might want to post it [here](http://stackoverflow.com/questions/tagged/amber-smalltalk). You may as well edit the [wiki](https://github.com/amber-smalltalk/amber/wiki/ToDo-list-application) page and ask questions there.

**Comment by herby** [comment]
Labelled as "won't fix" as this is a user question not an Amber issue. Maybe @amroto can even close it.

**Comment by hhzl** [comment]
@amroto, yes, please close the issue and move the discussion either to

1.
http://stackoverflow.com/questions/tagged/amber-smalltalk (for
questions about how to construct GUI, you have to provide an inital
attempt at a code snippet executing a task for it to be an acceptable
stackoverflow question)


2.
or to a wiki page https://github.com/amber-smalltalk/amber/wiki in
connection with questions on the mailing list (for constructing a GUI
example)

**Comment by hhzl** [comment]
@amroto

You may as well open an issue
[here](https://github.com/amber-smalltalk/amber-examples/issues) about
the construction of the example you suggested

**Comment by hhzl** [comment]
@amroto and use https://gitter.im/amber-smalltalk/amber
